File: sky/engine/core/painting/ImageFilter.cpp
// Copyright 2016 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "sky/engine/core/painting/ImageFilter.h"

#include "sky/engine/tonic/dart_args.h"
#include "sky/engine/tonic/dart_binding_macros.h"
#include "sky/engine/tonic/dart_converter.h"
#include "sky/engine/tonic/dart_library_natives.h"
#include "third_party/skia/include/effects/SkBlurImageFilter.h"
#include "third_party/skia/include/effects/SkImageSource.h"
#include "third_party/skia/include/effects/SkPictureImageFilter.h"

namespace blink {

static void ImageFilter_constructor(Dart_NativeArguments args) {
  DartCallConstructor(&ImageFilter::create, args);
}

IMPLEMENT_WRAPPERTYPEINFO(ui, ImageFilter);

#define FOR_EACH_BINDING(V) \
  V(ImageFilter, initImage) \
  V(ImageFilter, initPicture) \
  V(ImageFilter, initBlur)

FOR_EACH_BINDING(DART_NATIVE_CALLBACK)

void ImageFilter::RegisterNatives(DartLibraryNatives* natives) {
  natives->Register({
    { "ImageFilter_constructor", ImageFilter_constructor, 1, true },
FOR_EACH_BINDING(DART_REGISTER_NATIVE)
  });
}

scoped_refptr<ImageFilter> ImageFilter::create() {
  return new ImageFilter();
}

ImageFilter::ImageFilter() {
}

ImageFilter::~ImageFilter() {
}

void ImageFilter::initImage(CanvasImage* image) {
  filter_ = SkImageSource::Make(image->image());
}

void ImageFilter::initPicture(Picture* picture) {
  filter_ = SkPictureImageFilter::Make(picture->toSkia());
}

void ImageFilter::initBlur(double sigmaX, double sigmaY) {
  filter_ = SkBlurImageFilter::Make(sigmaX, sigmaY, nullptr);
}

} // namespace blink

File: engine/src/flutter/lib/web_ui/lib/src/engine/surface/image_filter.dart
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

// @dart = 2.6
part of engine;

/// A surface that applies an [imageFilter] to its children.
class PersistedImageFilter extends PersistedContainerSurface
    implements ui.ImageFilterEngineLayer {
  PersistedImageFilter(PersistedImageFilter oldLayer, this.filter) : super(oldLayer);

  final ui.ImageFilter filter;

  @override
  html.Element createElement() {
    return defaultCreateElement('flt-image-filter');
  }

  @override
  void apply() {
    rootElement.style.filter = _imageFilterToCss(filter);
  }

  @override
  void update(PersistedImageFilter oldSurface) {
    super.update(oldSurface);

    if (oldSurface.filter != filter) {
      apply();
    }
  }
}
